***DISCLAIMER (Read this carefully)***

Before you turn this assignment in, make sure everything runs as expected. First, restart the kernel (in the menubar, select Kernel
Restart) and then run all cells (in the menubar, select Cell
Run All). Do NOT add any cells to the notebook!

Do not forget to submit both the notebook AND the files in the data/ subfolder according to the CoC!
Make sure you fill in any place that says YOUR CODE HERE or YOUR ANSWER HERE , as well as your name and group below:

In [1]:
# These are the libraries you are allowed to use for this assignment
# If any of those libraries are not present in your jupyter instance, install them by using !pip install [library-name]!
!pip install xmltodict
import xmltodict
import csv
import json
!pip install pytest
import pytest
import requests
import codecs
import math
import re
import requests
import urllib.request
import pandas
import xml.etree.ElementTree as ET

# Assignment 2 (Group)
When carrying out a Data-Science project, screening and selecting appropriate data sources for the tasks at hand comes at the beginning. This assignment is about accessing and characterising potential data sources in teams of three. The teams have been randomly assigned. BEWARE! In Assignment 5, you will be asked to provide answers to those questions. Make sure that combining the two datasets makes sense from an analytical perspective!

-----
## Step 0 (2 points)

Find two data sets online (from one or several sources) that would be interesting to combine and create ***data citations*** as Python dictionaries. 

The data sets should fulfill the following requirements:

* Each data set must have a different file format (either CSV, XML, or JSON), please choose
  - one CSV file (dataset1) 
  - and one JSON or XML file (dataset2)

* The two datasets should not be two variations of each other (i.e. simply the same dataset for two different regions or timeframes or from the same source just in two different formats)
* Workable data-set sizes: The selected or extracted data sets should have thousands of entries (>= 1000), but not more than (<=) 10000 entries. Be "entries we mean rows or distinguishable key-value pairs). If larger, use an excerpt from the original data set. Justify in detail the extraction criteria in the markdown cell below and 
  1) add the code used for the extraction in the code cell or describe how you filtered the sample  
  2) make the extracted dataset also available at a downloadable URL (for instance in a Github repository, [here](https://raw.githubusercontent.com/AxelPolleres/simple_dataset_sharing_repo/main/test.csv)'s an example)
  3) name the new `resourceURL` in the data citation.
* You may start from (but you are not limited to) the resource collections hinted at [in the Unit 2 slides](https://datascience.ai.wu.ac.at/ws21/dataprocessing1/unit2.html#slide-53).

* Important: The use of datasets from kaggle.com and other curated collections of datasets with accompanying tutorials on processing and analysis (as highlighted to you in Unit 2) is **discouraged**. You are required to use **primary data sources**: This is mainly because we want you to work on data sets that have not been processed with some analysis in mind, so that you show that you can handle (messy) data sets harvested on the brownfields of Data Science. Besides, such curated datasets have been repeatedly used in ready-made case and tutorial work, which makes it basically impossible for us to establish whether your submissions are genuine contributions of yours. There is one viable option: Work backwards from the Kaggle data set to the original data source, obtain updated data from there, and start from there.


* Please adhere to the CoC.

[Data citations](http://blogs.nature.com/scientificdata/2016/07/14/data-citations-at-scientific-data/) must contain the following details:
- creator: provider organisation / author(s) of the data set, e.g. "Zentralanstalt für Meteorologie und Geodynamik (ZAMG)"
- catalogName: Names of the data repository and/or the Open Data portal used, e.g. Open Data Österreich"
- catalogURL: URL of th repository / portal, e.g. "https://www.data.gv.at/"
- datasetID: (specific to the data repository), e.g. "https://www.data.gv.at/katalog/dataset/zamg_meteorologischemessdatenderzamg"
- resourceURL: a URL where the CSV, XML or JSON file can be downloaded, e.g. "https://www.football-data.co.uk/new/JPN.csv"
- pubYear: Dataset publication year, i.e. since when it is published, e.g. "2012"
- lastAccessed: when have you last accessed the dataset (i.e. datetime of accessing, obtaining a copy of the data set) in ISO Format? e.g. "2021-03-08T13:55:00"

One final note: as mentioned above, if you want to use a repository for your file download (e.g. github), you are allowed to do that. The most important part is that the URL can be accessed stably for each dataset you have chosen. 

Store the data citation in a dictionary for each of the datasets:

In [2]:
# We cut down the data, using a script that first delets the data which isnt in the correct time period (so that both datasets are in teh same time period) and then randomly selects 10000 samples. The code for both the CSV and JSON can be found in the respective Github Repos.
dataset1 = {
    "creator": "U.S. Department of Housing and Urban Development (HUD)",
    "catalogName": "Neighborhood Data for Social Change",
    "catalogURL": "https://la.myneighborhooddata.org/",
    "datasetID": "https://la.myneighborhooddata.org/2021/06/area-income-limits/",
    "resourceURL": "https://raw.githubusercontent.com/anjatheanja/Assignment-2/main/final_processed_data.csv",
    "pubYear": "2021",
    "lastAccessed": "2025-03-22T19:28:34",
}

dataset2 = {
    "creator": "Los Angeles Police Department",
    "catalogName": "Crime Data from 2010 to 2019",
    "catalogURL": "https://data.lacity.org",
    "datasetID": "https://data.lacity.org/Public-Safety/Crime-Data-from-2010-to-2019/63jg-8b9z/about_data",
    "resourceURL": "https://raw.githubusercontent.com/brechthaben/BBE/main/SBWL%20DataScience/DataProcessing/Assignments/Assignment2/processed_data.json",
    "pubYear": "2019",
    "lastAccessed": "18.03.2024"
}

In [3]:
import traceback
import sys
import os

assert type(dataset1) == dict
assert type(dataset2) == dict


## Please answer the questions in the cell below

Use the following structure for your answer below:

Data set 1

(Describe the source and the general content of the dataset and why you chose it)

Data set 2

(Describe the source and the general content of the dataset and why you chose it)

Project ideas

(Describe in your own words, which kind of tasks could be addressed by combining the selected data sets, esp. how the two data sets fit together and what complementary information they contain; Formulate a question that could be potentially answered by combining data from both datasets; how could the data sets be combined exactly? 250 words max. BEWARE! In Assignment 5, you will be asked to provide answers to those questions. Make sure that combining the two datasets makes sense from an analytical perspective!)

Data set 1
This was the second data set that was found. Since we were looking for something that could coorelate to the first dataset in a way - it was very hard to find something this tailored. It is not optimal, especially because of the regions but it does provide us with information about household income which can be a great factor for analysis.

Data set 2

We chose this dataset after a long search. We wanted something with crime data and then compare them to socioeconomic factors. This was one of the few datasets we found which was detailed enough and where we could find a dataset with those factors. The source is the Los Angeles Police Department. It originally included 2.1 million individual crime reports from 2010 to 2019 but also got cut down to 10000. The code for this can be found in the same directory on github.

Project ideas

We hope to investigate the crime rates for poorer and richer areas in Los Angeles. We should be able to map the regions of both onto eachother. This would involve a third dataset which links streets to a specific area indicator, like the postal code. 

Potential questions could be: 
What crime is most common in different income brackets?
Does crime happen more often in poorer neighborhoods?
Is there is a proportion of increase in crime rate vs income decrease?

------
## Step 1 - File Access (3 points)

Write a Python function `accessData` that takes the dataset dictionary created in step 0 as an input and returns an extended dictionary including following additions:

* Write code that accesses the dataset from its `resourceURL` using the python `requests` package:
 * detects whether it's and XML, CSV or JSON file by
     * checking whether the download URL **ends** with suffix "xml", "json", "csv" (in either upper- or lowercase)
     * checking whether the "Content-Type" HTTP header field contains information about the format, hinting on XML, JSON or CSV, i.e., check whether the substring XML, JSON or CSV appears in the "Content-Type" header in either upper- or lowercase. 
 * Detects the file size from the HTTP header (converted to KB) of each data set, clearly documenting your actions (e.g. through commented code).

The result of the code below should extend your dictionaries `dataset1` and `dataset2` with two keys named 
* `"detectedFormat"` (which has one of the following values: `"XML"`, `"JSON"`, `"CSV"`, or `"unknown"`, if nothing could be detected from checking the suffix or HTTP header, or if the information in both was inconsistent)
* and `"filesizeKB"` which contains the filesize in KB (Conversion should be done accordingly to decimal SI prefixes) from the number of bytes in the header-information. If there is no respective header information return 0.
* If the detected format is `"unknown"`, the expected filesize to be returned is also 0


In [4]:
def accessData(datadict):    
    
    #Fetching resource URL from dictionary
    URL = datadict.get("resourceURL", "")
    
    #initial format and size
    detectedFormat = ""
    filesizeKB = 0

    #Send Head request and save headers
    response = requests.head(URL)
    headers = response.headers

    # Checking download URL suffix -> either small or capital
    #for csv 
    if ".csv" in URL or ".CSV" in URL:
        detectedFormat = "CSV"
    #for json     
    elif ".json" in URL or ".JSON" in URL:
        detectedFormat = "JSON"
    #for xml 
    elif ".xml" in URL or ".XML" in URL:
        detectedFormat = "XML"

    #Checking if headers have hint suffix in Content-Type
    if detectedFormat == "" and "Content-Type" in headers:
        
        # Pick only Content-Type from headers
        contentType = headers["Content-Type"]
        
        #same as suffix from download URL
        #for csv
        if "csv" in contentType or "CSV" in contentType:
            detectedFormat = "CSV"
            
        #for json
        elif "json" in contentType or "JSON" in contentType:
            detectedFormat = "JSON"
        
        #for xml
        elif "xml" in contentType or "XML" in contentType:
            detectedFormat = "XML"

    #choose only content length from headers
    if "Content-Length" in headers:
        
        #fetch content length as integer and translate bytes to kb -> ( 1000 bytes = 1kB )
        filesizeKB = int(headers["Content-Length"]) / 1000 

    #add the variables to the dictionary
    datadict["filesizeKB"] = filesizeKB
    datadict["detectedFormat"] = detectedFormat
    
    return datadict  
# Process both datasets
accessData(dataset1)
accessData(dataset2)

print("Dataset 1:")
print(f"File size: {dataset1['filesizeKB']} KB")
print(f"Detected format: {dataset1['detectedFormat']}")

print("\nDataset 2:")
print(f"File size: {dataset2['filesizeKB']} KB")
print(f"Detected format: {dataset2['detectedFormat']}")

Dataset 1:
File size: 451.551 KB
Detected format: CSV

Dataset 2:
File size: 764.137 KB
Detected format: JSON


In [5]:
# Basic tests to see if your solution meets the foundational demands described in the task description
dataset1 = accessData(dataset1)
dataset2 = accessData(dataset2)
if dataset1["detectedFormat"] not in ["XML", "JSON", "CSV", "unknown"]:
    raise AssertionError
if dataset2["detectedFormat"] not in ["XML", "JSON", "CSV", "unknown"]:
    raise AssertionError

assert isinstance(dataset1["filesizeKB"],(int, float)) == True
assert isinstance(dataset2["filesizeKB"],(int, float)) == True

In [6]:
# Edit this cell or remove it, and you shall perish, meow! 😼⚡️

In [7]:
# I know you are reading this, there is no help here, no hope. Also, do not delete me please :3

## Please answer the questions in the cell below

#### Please explain your findings, using the following structure for your answer below (in "other remarks" you can explain, for instance, why you think your code did not detect the correct format, if needed)

Data set 1

(format, size, other remarks)

Data set 2

(format, size, other remarks)

Data set 1 

File size: 451.551 KB
Detected format: CSV


Data set 2
File size: 764.137 KB
Detected format: JSON

This is a JSON which is 764137KB large

-----
## Step 2  (5 points) - Format Validation

Establish that the two data files obtained are well-formed according to the detected data format (CSV, JSON, or XML). That is, the syntax used is valid according to accepted syntax definitions. Are there any violations of well-formedness?


Proceed as follows (for each data file, in turn): according to the "suspected" data format from Step 1:

  1. Use an _online validator_ for CSV, XML, and JSON, respectively, to confirm whether the files you downloaded in Step 1 are well-formed for the respective file format, document your findings and modify the file as described: 

   a. **Case 1**: no well-formedness errors were detected: 
    * Generally describe at least 3 well-formedness checks that your data sets, depending on its "suspected" format (against the background knowledge of Unit 2) should fulfill;
    * Store a local copy of the file called `data_notebook-[notebook-nr.]_[name].[file extension]` in the `data/` subfolder
    * Create another local copy of your data file called `data_notebook-[notebook-nr.]_[name]-invalid.[file extension]` and introduce a selected well-formedness violation (one occurrence) therein;
    * document that the online validator you used finds the error you introduced

   b. **Case 2**: well-formedness errors occurred:
    * Document the occurrences by printing out the error message and describe the types of well-formedness violation that were reported to you.
    * Store a local copy called `data_notebook-[notebook-nr.]_[name]-invalid.[file extension]`  in the `data/ subfolder`
    * Create another local copy called `data_notebook-[notebook-nr.]_[name].[file extension]`, of your data file that fixes the well-formedness violations therein manually.  
    
**Please note that the datasets in the `data/` subfolder are for documentation only. Do not access those for subsequent steps!**
    

  2. Write a Python function `parseFile(datadict, format)` that that accesses the dataset from its `resourceURL`. The dataset should then be checked accordingly the given parser for the parameter `format` to check the following:
     * CSV: Returns `True`, if a consistent delimiter out of `",",";","\t"` can be detected, such that each row has the same (> 1) number of elements, otherwise False
     * JSON: Returns `True` if the file can be parsed with the `json` package, catching any parsing exceptions.
     * XML: Returns `True` if the file can be parsed with the `xmltodict` package, catching any parsing exceptions.
     * Returns `False` if any other format is supplied by the parameter.
     
In order to handle parsing exceptions and errors from the used packages, you can use [catching exceptions](https://docs.python.org/3/tutorial/errors.html), such that the program does not simply fail to check whether the file is parseable as the format specified in `format`    

## Please answer the questions in this cell

### Use the following structure for your answer in the cell below to document **Step 2.1**:

***Data set 1***

*(validator used, validation results, describe the modification to fix the file or to create an invalid version of it)*

***Data set 2***

*(validator used, validation results, describe the modification to fix the file or to create an invalid version of it)*


Data set 1
We used an online CSV validator. No errors were detected. To create an invalid version, we could have an inconsistent ammount of collumns. Adding special characters or creating inconsistent ammount of commas will also cause the file to be invalid. Well formedness can come from consistent collumns, consistent delimitors and consistent field quotation.



Data set 2

The JSON validator had no issues either. To create an invalid versison we could delete braces, use brackets and braces interchangeable or dont use quotation marks. Similarly the well formedness comes from consistent braces, delimiters and quotations.

In [15]:
import requests
import csv
import json
import xmltodict
import io  # Used to convert string content to a file-like object for CSV parsing

def parseFile(datadict, format):
   
     # Get the URL from the dictionary. Use an empty string if not found.
    resource_url = datadict.get("resourceURL", "")
    
    try:
        # Download the file content from the resourceURL
        response = requests.get(resource_url)
        response.raise_for_status()  # Raise an exception for HTTP errors
        content = response.text
    except Exception as e:
        
        # In case of any error, return False.
        return False
    
    # Convert the format to uppercase to avoid case issues.
    format_upper = format.upper()
    
    if format_upper == "CSV":
        # For CSV, check for a consistent delimiter among [",", ";", "\t"]
        possible_delimiters = [",", ";", "\t"]
        consistent = False
        
        # Try each possible delimiter.
        for delim in possible_delimiters:
            
            # Create a file-like object from the downloaded text.
            f = io.StringIO(content)
            reader = csv.reader(f, delimiter=delim)
            rows = list(reader)
            
            if not rows:
                
             # If no rows are found, try the next delimiter.
                continue
                
            # Use the first row to determine the expected number of columns.
            expected_columns = len(rows[0])
            
            # Only consider valid if there is more than one column
            if expected_columns <= 1:
                continue
                
            # Check that every row has the same number of columns
            if all(len(row) == expected_columns for row in rows):
                consistent = True
                break # A good delimiter is found, so stop the loop.
        return consistent
    
    elif format_upper == "JSON":
        try:
            # Try to parse the file content using the json package
            json.loads(content)
            return True
        except Exception as e:
            return False
    
    elif format_upper == "XML":
        try:
            # Try to parse the file content using the xmltodict package
            xmltodict.parse(content)
            return True
        except Exception as e:
            return False
    else:
        # If the format is not recognized, return False.
        return False


In [16]:
# This checks if your code returns true twice for either xml, json or csv for your datasets specifically.

assert([parseFile(dataset1, "XML"),
    parseFile(dataset1, "JSON"),
    parseFile(dataset1, "CSV"),
    parseFile(dataset2, "XML"),
    parseFile(dataset2, "JSON"),
    parseFile(dataset2, "CSV")].count(True)) == 2

In [10]:
# There will be consequences if you remove this cell, or even move this cell. Be wary of that <3

-----
## Step 3 - Content analysis (5 points)

Similar to the Python function `parseFile(datadict,format)` above, now create a new Python function `describeFile(datadict)` that analyses the given file according to the respective format detected in Step 1 and returns a dictionary containing the following information:

* for CSV files: number of numeric (float) columns, number of rows, column index (from 0 to n) and row index of the entry which contains the largrst float value which is a round number (modulo 2 with a remainder of 0). Ensure to try to transform any entry to a numeric type to ensure no number is accidentely seen as a string. The resulting dictionary should have the following form:

    ```
    { "numberOfNumericColumns:"  ...,
       "numberOfRows":  ... ,
       "largestFloat" : ... }
    ```

* for JSON files: number of distinct attribute names, nesting depth, length of the longest list appearing in an attribute value. That is, the resulting dictionary should have the following form:

    ```
    { "numberOfAttributes:" ... ,
      "nestingDepth":  ... ,
      "longestListLength" : ... }
     ```

  Here the `longestListLength` should be set to 0 if no list appears. [Nesting depth](https://www.tutorialspoint.com/find-depth-of-a-dictionary-in-python) is defined as follows: 
   * a flat list of atomic values has depth 0, a flat JSON object with only atomic attribute values has depth 1. 
   * a JSON attribute with another object as value (or another object as member of a list value!) increases the depth by 1
   * and so on.


* for XML files: number of different element and attribute a names (i.e. the sum of both), nesting depth, maximum numeric value in the dataset. That is, the resulting dictionary should have the following form:

    ```
    { "numberOfElementsAttributes:" ... ,
      "nestingDepth":  ... ,
      "longestString" : ... }
     ```

  Here the `longestString` should be set to "" if there are no String values present. Nesting depth is defined as the nesting depth of elements.
  
For files that cannot be parsed with respective given format, the function should simply return an empty dictionary (`{}`).

##### Please ensure to keep the order of the elements as specified above! The test cases assume that your result has the same structure as the ones specified above. Basically this means for the csv e.g. that the first position should ALWAYS be the numberOfNumericColumns!

In [11]:
def describeFile(datadict):

    #getting all the needed info from the dictionary
    URL = datadict.get("resourceURL", "")
    detectedFormat = datadict.get("detectedFormat", "")

    # Basic check to see if the URL or format is missing
    if not URL or not detectedFormat:
        return {}

    #if the format -> CSV :
    if detectedFormat == "CSV":
        try:
            #download csv from url as a string
            # added timeout to be kind to the server :)
            response = requests.get(URL, timeout=10)
            response.raise_for_status() # Raise an exception for bad status codes
            content = response.text

            # now we Check for empty content after stripping whitespace
            if not content.strip():
                return {}

            #remove spaces & split into rows
            #Using splitlines() for better cross-platform line ending support
            rows = content.strip().splitlines()

            # not needed here, but for futureproofing we check if the file is empty or only a header
            if not rows:
                return {}
            header_row = rows[0]
            data_rows = rows[1:]

            # here we check for the case, that there are no data rows
            if not data_rows:
                 return {
                    "numberOfNumericColumns": 0,
                    "numberOfRows": 0,
                    # we use (-1, -1) to indicate not found, as requested index starts from 0
                    "largestFloat": (-1, -1)
                }

            #try these delimeters for file separation
            potential_delimiters = [",",";","\t", "|"]
            best_delimiter = None
            splitRows = []
            num_cols_header = 0 #Store header column count for consistency check

            # here we try to detec the delimiter being used
            for delimiter in potential_delimiters:
                # Check delimiter on header first
                header_parts = header_row.split(delimiter)
                current_num_cols = len(header_parts)
                if current_num_cols <= 1: # Skip if delimiter doesn't split header meaningfully
                    continue

                temp_splitRows = []
                consistent = True
                for row_str in data_rows:
                    parts = row_str.split(delimiter)
                    # Check if row has more columns than header - this indicates wrong delimiter
                    if len(parts) > current_num_cols:
                        consistent = False
                        break
                    # Pad rows that are shorter than the header
                    parts.extend([''] * (current_num_cols - len(parts)))
                    temp_splitRows.append(parts)

                if consistent:
                    best_delimiter = delimiter
                    splitRows = temp_splitRows
                    num_cols_header = current_num_cols
                    break # Found a consistent delimiter

            # now for the case, that our delimiters arent enough: If no consistent delimiter worked based on header length, return empty
            if best_delimiter is None:
                # Maybe add a fallback here if needed, but for now stick to consistency check
                return {}


            # Get number of rows excluding header
            numberOfRows = len(splitRows)
            # num_cols = len(splitRows[0]) if numberOfRows > 0 else num_cols_header
            num_cols = num_cols_header # Use the number of columns derived from the header and consistent rows

            numericColumnIndices = []

            # Go through each row by the index based on rows legth 
            for i in range(num_cols):
                #assume the column is numeric
                isNumeric = True
                # Loop through the each row, skip header and only include non-empty rows, convert into float if can
                for row in splitRows:
                    # now we Check index bounds explicitly although padding should handle it
                    if i < len(row):
                        value_str = row[i].strip() # here we Strip whitespace from value
                        if value_str != "": # and now we test non-empty strings to really be bulletproof
                            try:
                                float(value_str)
                            except (ValueError, TypeError): # Catch TypeError as well
                                isNumeric = False
                                break # Once a non-numeric is found, stop checking this column

                if isNumeric:
                    # here we Store the index of the numeric column
                    numericColumnIndices.append(i)

            #how many numeric columns were found
            numberOfNumericColumns = len(numericColumnIndices)

            #looking for the largest even float ( + the position)
            largestValue = None
            # explanation: (columnIndex, rowIndex)
            # we Initialize position to indicate "not found" clearly. see above for context
            position = (-1, -1)

            #loop through every row and cell -> looking for even float
            for rowIndex, row in enumerate(splitRows):
                #only in confirmed numeric columns -> avoids string based IDs and all of that
                for columnIndex in numericColumnIndices:
                    try:
                        #get the value from the num column
                        value = row[columnIndex]
                        # now we again Strip whitespace before converting
                        value_str = value.strip()
                        if value_str: # Check if string is not empty after stripping
                             #if the num can be divided by 2 -> even, number is numeric (converted into float)
                            number = float(value_str)
                            # Check if it's effectively an integer (no fractional part) and even
                            # Also check number == number to exclude NaN/inf which cause errors with modulo
                            if number == number and number % 1 == 0 and number % 2 == 0:
                                # Store the found current largest value and get its position
                                if largestValue is None or number > largestValue:
                                    largestValue = number
                                    # keep in mind: rowIndex is 0-based index within splitRows (data rows)
                                    position = (columnIndex, rowIndex)
                    except (ValueError, TypeError): # Catch TypeError again 
                        #ignore if it is not a number
                        continue

            # here we want to ensure the key order matches the required specification
            return {
                "numberOfNumericColumns": numberOfNumericColumns,
                "numberOfRows": numberOfRows,
                "largestFloat": position
            }

        except requests.exceptions.RequestException as e: # we try to catch specific request errors
            print(f"TODO - CHANGE COMMENT: CSV Request failed: {e}") # helps with debugging
            return {}
        except Exception as e: # and now we (dont) want to Catch other potential processing errors
            print(f"TODO - CHANGE COMMENT: CSV Processing failed: {e}") # helps with debugginh
            return {}

    #if format -> JSON then:
    elif detectedFormat == "JSON":
        try:
            #fetching JSON data from URL
            # again a timeout to ensure we dont upset the robot overlords
            response = requests.get(URL, timeout=10)
            response.raise_for_status()
            # we again try to Handle potential empty response body before JSON parsing
            if not response.content:
                 return {}
            data = response.json()

            #counting the unique attribute names
            def collectAttributes(obj, names_set):
                 #if we have a dictionary -> (example: {"name": "John", "age": 25})
                if isinstance(obj, dict):
                    #add keys from current dict
                    for key in obj.keys():
                        names_set.add(key)
                    #explore the values (check for nested dict/ or lists)
                    for value in obj.values():
                        collectAttributes(value, names_set)
                #if list -> go through every item & find more keys
                elif isinstance(obj, list):
                    for item in obj:
                        collectAttributes(item, names_set)

            attributeNames = set() #lets give our attributes a nice place to live in <3
            collectAttributes(data, attributeNames) 


            #function for finding the longest nesting depth
            def findDepth(obj):
                #if obj is a dictionary -> {"name": "John", "age": 25}
                if isinstance(obj, dict):
                    # the base depth for an object is 1
                    max_child_depth = 0
                    # If dict is empty, depth is 1. Otherwise, 1 + max depth of children.
                    if not obj:
                        return 1
                    #go through each value in the dictionary
                    for value in obj.values():
                        # now we recursively find depth of each value
                        depth = findDepth(value)
                        if depth > max_child_depth:
                            max_child_depth = depth
                    # this gives us the max depth +1 because of the base depth
                    return 1 + max_child_depth

                #if obj is a list
                elif isinstance(obj, list):
                    # the Base depth for a list is 0, depth is max depth of items.
                    max_item_depth = 0
                    # If list is empty, depth is 0. Otherwise, max depth of items.
                    if not obj:
                        return 0
                    # Go through each item in the list
                    for item in obj:
                         # we again find the depth of each item 
                        depth = findDepth(item)
                        if depth > max_item_depth:
                            max_item_depth = depth
                    # and now we Return the max depth found among items (list itself doesn't add depth per definition)
                    return max_item_depth

                #if obj is a number or a string
                else:
                    # Atomic values have depth 0 according to definition
                    return 0

            #function to find longest list anywhere in the JSON file
            def findLongestList(obj):
                longest = 0

                # If we have a dictionary -> checking for lists / nested structures
                if isinstance(obj, dict):
                    for value in obj.values():
                        if isinstance(value, list):
                             # we want to Compare current list length with longest found so far
                            longest = max(longest, len(value))
                        # and now we Recurse into nested dicts/lists
                        if isinstance(value, (dict, list)):
                            longest = max(longest, findLongestList(value))

                #if we have a list -> check length / nested structures
                elif isinstance(obj, list):
                     # we also Consider the length of the current list itself
                    longest = max(longest, len(obj))
                    for item in obj:
                         # again Recurse! into nested dicts or lists (again)
                        if isinstance(item, (dict, list)):
                            longest = max(longest, findLongestList(item))

                return longest

            #create result dictionary
            result = {
                "numberOfAttributes": len(attributeNames), 
                "nestingDepth": findDepth(data),
                "longestListLength": findLongestList(data)
            }

            return result

        except requests.exceptions.RequestException as e: # again error handling to make this even more powerful (unicode U+1F4AA for the emoji enjoyers;) )
            print(f"TODO - CHANGE COMMENT: JSON Request failed: {e}") 
            return {}
        except json.JSONDecodeError as e: 
            print(f"TODO - CHANGE COMMENT: JSON Parsing failed: {e}") 
            return {}
        except Exception as e: 
            print(f"TODO - CHANGE COMMENT: JSON Processing failed: {e}") 
            return {}

    #if the file is xml
    elif detectedFormat == "XML":
        # now we Implement the XML analysis, eventhough we dont have an XML. we werent quite sure if this is necessary but dont want points deducted
        try:
            # now we Fetch XML data from URL like a cowboy
            response = requests.get(URL, timeout=10)
            response.raise_for_status()
            # we Use response.content, because Dr. Google said it handles encoding better ¯\_(ツ)_/¯
            xml_content = response.content
            # now we want to handle empty responses again
            if not xml_content.strip():
                 return {}

            # finally we get to the processing: we parse the XML content
            root = ET.fromstring(xml_content)

            # our little helper function will collect element and attribute names
            def collect_xml_names(element, names_set):
                if element is None:
                    return
                # we add the tag name of the current element
                names_set.add(element.tag)
                # then we add all attribute names of the current element
                for attr_name in element.attrib:
                    names_set.add(attr_name)
                # now we recursively call for all child elements
                for child in element:
                    collect_xml_names(child, names_set)

            # this little guy will calculate the XML nesting depth 
            def get_xml_depth(element):
                # in XML an element itself has depth 1
                if not list(element): # Check if the element has any child elements
                    return 1
                else:
                    # and now we calculate depth recursively for children
                    max_child_depth = 0
                    for child in element:
                       child_depth = get_xml_depth(child)
                       if child_depth > max_child_depth:
                           max_child_depth = child_depth
                    # the depth is 1 (for current element) + max depth of children
                    return 1 + max_child_depth

            # the last helper function to find the longest string in text or attributes
            def find_longest_xml_string(element, current_longest):
                 # Tfirst we check the element's text content
                 if element.text:
                     text = element.text.strip()
                     if len(text) > len(current_longest):
                         current_longest = text

                 # then we check the element's attribute values
                 for attr_value in element.attrib.values():
                     # Ensure value is a string before checking length
                     if isinstance(attr_value, str):
                         if len(attr_value) > len(current_longest):
                             current_longest = attr_value

                 # and now we recursively check children, passing the longest found so far
                 for child in element:
                     current_longest = find_longest_xml_string(child, current_longest)

                 return current_longest # Return the longest string found in this branch


            # Home stretch! We Calculate the required values
            xml_names = set()
            collect_xml_names(root, xml_names)
            numberOfElementsAttributes = len(xml_names)

            # then we calculate depth starting from the root
            nestingDepth = get_xml_depth(root) if root is not None else 0

            # and now we will find the longest string, initialize with empty string
            longestString = find_longest_xml_string(root, "") if root is not None else ""

            # done at last! we create the result dictionary in the specified order
            return {
                "numberOfElementsAttributes": numberOfElementsAttributes,
                "nestingDepth": nestingDepth,
                "longestString": longestString
            }

        except requests.exceptions.RequestException as e: # One last time - to help with errors and to make this as strong as schwarzenegger
            print(f"TODO - CHANGE COMMENT: XML Request failed: {e}") 
            return {}
        except ET.ParseError as e: 
            print(f"TODO - CHANGE COMMENT: XML Parsing failed: {e}") 
            return {}
        except Exception as e: 
            print(f"TODO - CHANGE COMMENT: XML Processing failed: {e}") 
            return {}

    # now to deal with files that arent CSV, JSON or XML!
    else:
        return {} 
results1 = describeFile(dataset1)
results2 = describeFile(dataset2)



In [12]:
assert(len(describeFile(dataset1))) == 3
assert(len(describeFile(dataset2))) == 3

# Check your output
print(describeFile(dataset1))
print(describeFile(dataset2))

{'numberOfNumericColumns': 13, 'numberOfRows': 10000, 'largestFloat': (1, 529)}
{'numberOfAttributes': 27, 'nestingDepth': 1, 'longestListLength': 10000}


In [13]:
# Edit this cell or remove it, and you shall perish, meow! 😼⚡️

In [14]:
# Edit this cell or remove it, and you shall perish, meow! 😼⚡️

## Please answer the questions

***Final check:***

Be sure to cross-check the results by
1) manually inspecting your chosen dataset and
2) comparing the results for plausibility against the results of your code... 

Describe your findings and use the following structure for your answer below:

*Data set 1*

(number and types of items etc., describe your findings)

*Data set 2*

(number and types of items etc., describe your findings)

Data set 1

This dataset contains 13 numeric columns , therefore it is mainly quantitative. It was artificially cut down to 10000 rows previously and the code was able to detect that. The largest even float that was found is in column 1 in row 529. I manually checked the data to make sure it is the correct output.

Data set 2

Based on a careful visual inspection, the dataset comprises 27 columns, exhibits a nesting depth of 1, and contains exactly 10,000 artificially generated rows, confirming the initial analysis. Additionally, this manual verification directly addresses the question by ensuring that our code's output is both plausible and accurate. This data was also checked and matches the output
